catalog/core/catalog-core-standardframework/src/main/java/ddf/catalog/impl/action/SourceActionProviderImpl.java

```diff
@@ -0,0 +1,120 @@
+/**
+ * Copyright (c) Codice Foundation
+ *
+ * <p>This is free software: you can redistribute it and/or modify it under the terms of the GNU
+ * Lesser General Public License as published by the Free Software Foundation, either version 3 of
+ * the License, or any later version.
+ *
+ * <p>This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Lesser General Public License for more details. A copy of the GNU Lesser General Public
+ * License is distributed along with this program and can be found at
+ * <http://www.gnu.org/licenses/lgpl.html>.
+ */
+package ddf.catalog.impl.action;
+
+import ddf.action.Action;
+import ddf.action.ActionProvider;
+import ddf.action.impl.ActionImpl;
+import ddf.catalog.source.Source;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.util.Collections;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import org.apache.commons.lang.StringUtils;
+import org.apache.commons.lang3.tuple.ImmutablePair;
+import org.apache.commons.lang3.tuple.Pair;
+import org.boon.json.JsonFactory;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * Convert a JSON string to a map of source identifiers to action URLs.
+ */
+public class SourceActionProviderImpl implements ActionProvider {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SourceActionProviderImpl.class);
+
+  private static final String TITLE = "title";
+
+  private static final String URL = "url";
+
+  private Map<String, Action> sourceIdToAction;
+
+  private String actionProviderId;
+
+  public SourceActionProviderImpl(String actionProviderId) {
+    this.actionProviderId = actionProviderId;
+  }
+
+  @Override
+  synchronized public <T> Action getAction(T subject) {
```

Comment:
Would making this `<T extends Source>` remove the need for `instanceof` and casting?

Reply:
With `<T extends Source>` the method fails to satisfy the interface.